extensions/amp-base-carousel/1.0/arrow.js

```diff
@@ -25,17 +25,17 @@ export function Arrow({
   advance,
   by,
   customArrow = <DefaultArrow by={by} />,
-  disabled,
+  checkDisabled,
   outsetArrows,
   rtl,
 }) {
   const {
     'disabled': customDisabled,
     'onClick': onCustomClick,
   } = customArrow.props;
-  const isDisabled = disabled || customDisabled;
+  const disabled = checkDisabled() || customDisabled;
```

Comment:
A boolean prop would be preferred to a function that has to be called to get the actual value. This is because an arrow component wouldn't know when this function should be called. In this case it's called only at re-rendering, but someone should trigger the rerendering here. In which case, why not just pass a boolean? I think there are exception to this, especially when a context is used. But this case seems simpler?

Reply:
When it is called at the time of being passed into the arrow, it's not yet updated. Because we update the scroll position in a `useLayoutEffect`, we want to grab it sometime after the `BaseCarousel` renders and before it paints. This is the line we need to make sure runs first, since we are now measuring based on scroll position:
https://github.com/ampproject/amphtml/blob/604eb40832bc059fc180120da94b4844089f1654/extensions/amp-base-carousel/1.0/scroller.js#L151-L170

I had originally done `const [disabled, setDisabled] = useState(false)` and `useLayoutEffect(() => { setDisabled(checkDisabled()) }` but it appears that calling it directly in this component's render was correctly synchronized. 

But calling it at the time the `BaseCarousel` is rendered is not.

Reply:
We could pass `disabledRef={disabledRef}` to get the correct `disabledRef.current` value, but that to me felt just as different.

Reply:
But can we trust that "was correctly synchronized enough" is structurally stable enough?

Reply:
This relies on the assumption that effects should always run in the same order. Because we want `Arrow` to check its `disabled` state _after_ the scroll position is updated (`Scroller useLayoutEffect`), shouldn't it follow that if it does this once (manually verified), it should always do this where it makes a difference?

Reply:
But it also relies on the fact that `useLayoutEffect` runs before the arrow's render function? Or am I misunderstanding it?

Reply:
Yes that's right. Can that occur in arbitrary order?

Reply:
Well. I'd expect all renders to run at the same time, then all layout effects to run, and then renders to rerun if layout effects changed anything. So I'm still not quite sure how it works even now.

Reply:
Okay, I realized that one way to reduce the complexity here is to not try to wait for the scroll to occur. It is predictable what the target scroll position is, so we should base the disabled state off of that, and not what the scroll position of the container is at a certain point of time esp. if it is subject to change by other hooks. I restored the original `Arrow` logic and updated the disable fn instead in [838c696](https://github.com/ampproject/amphtml/pull/31765/commits/838c6968c8ef9d2ae0a91dc3f018b224b2383bec), ptal.

Reply:
Note there will have to be a slight adjustment to account for `snap-align=center`, but seeing as that is also somewhat buggy, I will fix those in the same PR.